src/protocols/bdx/BDXMessageUtils.h

```diff
@@ -0,0 +1,231 @@
+/*
+ *
+ *    Copyright (c) 2020 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      This file defines structures and utility methods for working with BDX
+ *      messages.
+ */
+
+#pragma once
+
+#include <support/CodeUtils.h>
+#include <system/SystemPacketBuffer.h>
+
+namespace chip {
+namespace BDX {
+
+enum ControlMode : uint8_t
+{
+    kNotSpecified  = 0x00,
+    kSenderDrive   = 0x10,
+    kReceiverDrive = 0x20,
+    kAsync         = 0x40,
+};
+
+/*
+ * A structure for representing a SendInit or ReceiveInit message (both contain
+ * identical parameters).
+ */
+struct TransferInit
+{
+    /**
+     * @brief
+     *  Pack (write) the message into a PacketBuffer.
+     *
+     * @param[out] aBuffer a PacketBuffer to use to write the message
+     *
+     * @return CHIP_ERROR Any error that occurs when trying to write to the PacketBuffer
+     */
+    CHIP_ERROR Pack(System::PacketBuffer & aBuffer);
+
+    /**
+     * @brief
+     *  Parse data from an PacketBuffer into a struct instance
+     *
+     * @param[in] aBuffer Pointer to a PacketBuffer containing the data.
+     * @param[out] aParsedMessage Reference to a struct instance where results will be stored.
+     *             Note that this struct will store pointers into the passed PacketBuffer, so it
+     *             is essential that the PacketBuffer is not modified or freed until after the
+     *             TransferInit struct is no longer needed.
+     *
+     * @return CHIP_ERROR Any error that occurs when trying to read the message
+     */
+    static CHIP_ERROR Parse(System::PacketBuffer & aBuffer, TransferInit & aParsedMessage);
```

Comment:
How about non-static and not have an out parameter?

Reply:
Should we force CHECK_RETURN_VALUE here?

Reply:
@bzbarsky-apple suggested a Constructor with an out parameter, but we decided that was not as clean.

The difference would be:
```
TransferInit msg;
err = msg.Parse(buf);
```
versus
```
TransferInit msg;
err = TransferInit::Parse(buf, msg);
```

Honestly, I did it this way because it's also how it was implemented in Weave. But either way, the caller has to instantiate the struct and then call Parse(). I guess it's cleaner to remove the `static`

Reply:
What's the criteria for forcing `CHECK_RETURN_VALUE`? I would think both `Pack()` and `Parse()` should always have the return value checked

Reply:
My rule of thumb would be "Anything returning CHIP_ERROR should have CHECK_RETURN_VALUE" since we never want to miss errors. Leads to a lot of boilerplate though, so feel free to not actually do it.

I prefer the non-static since I don't repeat as often the class name.